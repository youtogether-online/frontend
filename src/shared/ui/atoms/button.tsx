import { ButtonHTMLAttributes, forwardRef, ReactNode } from 'react'
import { styled } from '@/shared/config/stitches/stitches.config'
import { Text } from '@/shared/ui'
import { Loader } from './loader'

type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  variant?: 'solid' | 'text' | 'outlined' | 'icon' | 'link'
  theme?: 'primary' | 'danger' | 'success' | 'warning'
  size?: 'small' | 'medium' | 'large' | 'full'
  icon?: ReactNode
  pending?: boolean
  children?: ReactNode
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      variant = 'solid',
      pending,
      theme = 'primary',
      size = 'medium',
      icon,
      children,
      ...buttonProps
    },
    ref
  ) => {
    return (
      <ButtonStyled
        ref={ref}
        variant={variant}
        theme={theme}
        pending={pending}
        disabled={pending}
        size={size}
        {...buttonProps}
      >
        {!pending && icon}
        {!pending && children && <Text variant="button">{children}</Text>}
        {pending && <Loader />}
      </ButtonStyled>
    )
  }
)

const ButtonStyled = styled('button', {
  'transition': 'background 200ms',
  'outline': 'none',
  'border': 'none',
  'background': 'none',
  'cursor': 'pointer',
  'height': '36px',
  'position': 'relative',

  '& span': {
    fontSize: '$button',
  },

  'variants': {
    variant: {
      solid: {
        padding: '8px 16px',
        borderRadius: '$tertiary',
      },
      link: {
        height: 'auto',
        padding: 0,
        width: 'fit-content',
      },
      icon: {
        'padding': '2px',
        'borderRadius': '$full',
        'width': 'min-content',

        '&:hover': {
          backgroundColor: '$backgroundTextHover',
        },
      },
      text: {
        'background': 'none',
        'padding': '8px',
        'borderRadius': '$tertiary',

        '&:hover': {
          backgroundColor: '$backgroundTextHover',
        },
      },
      outlined: {
        'borderRadius': '$tertiary',
        'border': '1px solid $borderPrimary',

        '&:hover': {
          backgroundColor: '$backgroundInput',
        },
        '& svg': {
          display: 'block',
          position: 'absolute',
          left: '10px',
          top: '50%',
          transform: 'translateY(-50%)',
        },
      },
    },
    disabled: {
      true: {
        background: '$backgroundDisabled',
      },
    },
    size: {
      small: {
        width: '77px',
      },
      medium: {
        width: '80px',
      },
      large: {
        width: '93px',
      },
      full: {
        width: '100%',
      },
    },
    pending: {
      true: {
        background: '$backgroundDisabled',
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
      },
    },
    theme: {
      primary: {},
      danger: {},
      success: {},
      warning: {},
    },
  },

  'compoundVariants': [
    {
      variant: 'solid',
      theme: 'primary',
      css: {
        'backgroundColor': '$primary',
        'color': '$textWhite',
        '&:hover': {
          backgroundColor: '$primaryHover',
        },
      },
    },
  ],
})
