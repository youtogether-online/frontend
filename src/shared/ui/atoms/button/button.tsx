import { cva, type VariantProps } from "cva";
import { type ButtonHTMLAttributes, type ReactNode } from "react";
import { tw } from "typewind";

const button = cva(
  [tw.transition_colors, tw.border, tw.border_solid, tw.rounded_md, tw.cursor_pointer],
  {
    variants: {
      variant: {
        primary: [
          tw.bg_buttonPrimaryBg,
          tw.border_buttonPrimaryBorder,
          tw.text_buttonPrimaryText,
          tw.hover(tw.bg_buttonPrimaryHoverBg.border_buttonPrimaryBorder),
          tw.active(tw.border_buttonPrimaryBorder.bg_buttonPrimaryBg),
          tw.disabled(
            tw.bg_buttonPrimaryDisabledBg.text_buttonPrimaryDisabledText
              .border_buttonPrimaryDisabledBorder,
          ),
        ],
        secondary: [
          tw.bg_buttonBg,
          tw.text_buttonText,
          tw.border_buttonBorder,
          tw.hover(tw.border_buttonHoverBorder.bg_buttonHoverBg),
          tw.active(tw.bg_buttonBg.text_buttonText.border_buttonBorder),
          tw.disabled(tw.text_fgMuted),
        ],
        outline: [
          tw.bg_buttonBg,
          tw.text_buttonOutlineText,
          tw.border_buttonBorder,
          tw.hover(
            tw.bg_buttonOutlineHoverBg.border_buttonOutlineHoverBorder.text_buttonOutlineHoverText,
          ),
          tw.active(
            tw.bg_buttonOutlineSelectedBg.text_buttonOutlineSelectedText
              .border_buttonOutlineSelectedBorder,
          ),
          tw.disabled(tw.bg_buttonOutlineDisabledBg.text_buttonOutlineDisabledText),
        ],
        danger: [
          tw.bg_buttonBg,
          tw.border_buttonBorder,
          tw.text_buttonDangerText,
          tw.hover(
            tw.text_buttonDangerHoverText.border_buttonDangerHoverBorder.bg_buttonDangerHoverBg,
          ),
          tw.active(
            tw.text_buttonDangerSelectedText.border_buttonDangerSelectedBorder
              .bg_buttonDangerSelectedBg,
          ),
          tw.disabled(tw.bg_buttonDangerDisabledBg.text_buttonDangerDisabledText),
        ],
      },
      size: {
        sm: [tw.text_sm, tw.h_7, tw.px_2],
        md: [tw.text_md, tw.h_8, tw.px_3],
        lg: [tw.text_md, tw.h_10, tw.px_3],
      },
      disabled: {
        true: [tw.disabled(tw.pointer_events_none)],
      },
      block: {
        true: [tw.w_full],
      },
    },
    defaultVariants: {
      variant: "secondary",
      size: "md",
      block: false,
      disabled: false,
    },
  },
);

type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> &
  VariantProps<typeof button> & {
    leadingIcon?: ReactNode;
    trailingIcon?: ReactNode;
    children: ReactNode;
  };

export const Button = ({
  variant,
  block,
  size,
  disabled,
  leadingIcon,
  trailingIcon,
  children,
  ...buttonProps
}: ButtonProps) => {
  return (
    <button
      {...buttonProps}
      disabled={disabled}
      className={button({ variant, block, size, disabled })}
    >
      <span className={tw.flex.h_full.items_center.overflow_hidden.whitespace_nowrap}>
        {children}
      </span>
    </button>
  );
};
