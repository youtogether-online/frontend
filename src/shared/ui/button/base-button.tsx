import { cva, type VariantProps } from "cva";
import { tw } from "typewind";

import { type ButtonProps } from "./types";

export const baseButton = cva(
  [tw.transition_colors.border.border_solid.rounded_md.cursor_pointer.flex.gap_2.items_center],
  {
    variants: {
      variant: {
        primary: [
          tw.bg_buttonPrimaryBg.border_buttonPrimaryBorder.text_buttonPrimaryText,
          tw.hover(tw.bg_buttonPrimaryHoverBg.border_buttonPrimaryBorder),
          tw.active(tw.border_buttonPrimaryBorder.bg_buttonPrimaryBg),
          tw.disabled(
            tw.bg_buttonPrimaryDisabledBg.text_buttonPrimaryDisabledText
              .border_buttonPrimaryDisabledBorder,
          ),
        ],
        secondary: [
          tw.bg_buttonBg.text_buttonText.border_buttonBorder,
          tw.hover(tw.border_buttonHoverBorder.bg_buttonHoverBg),
          tw.active(tw.bg_buttonBg.text_buttonText.border_buttonBorder),
          tw.disabled(tw.text_fgMuted),
        ],
        outline: [
          tw.bg_buttonBg.text_buttonOutlineText.border_buttonBorder,
          tw.hover(
            tw.bg_buttonOutlineHoverBg.border_buttonOutlineHoverBorder.text_buttonOutlineHoverText,
          ),
          tw.active(
            tw.bg_buttonOutlineSelectedBg.text_buttonOutlineSelectedText
              .border_buttonOutlineSelectedBorder,
          ),
          tw.disabled(tw.bg_buttonOutlineDisabledBg.text_buttonOutlineDisabledText),
        ],
        danger: [
          tw.bg_buttonBg.border_buttonBorder.text_buttonDangerText,
          tw.hover(
            tw.text_buttonDangerHoverText.border_buttonDangerHoverBorder.bg_buttonDangerHoverBg,
          ),
          tw.active(
            tw.text_buttonDangerSelectedText.border_buttonDangerSelectedBorder
              .bg_buttonDangerSelectedBg,
          ),
          tw.disabled(tw.bg_buttonDangerDisabledBg.text_buttonDangerDisabledText),
        ],
        invisible: [
          tw.text_accentFg.border_none,
          tw.hover(tw.bg_buttonInvisibleHoverBg),
          tw.active(tw.bg_buttonInvisibleSelectedBg),
        ],
      },
      size: {
        sm: [tw.text_sm.h_7.px_2],
        md: [tw.text_md.h_8.px_3],
        lg: [tw.text_md.h_10.px_3],
      },
      disabled: {
        true: [tw.disabled(tw.pointer_events_none)],
      },
      block: {
        true: [tw.w_full],
      },
      alignContent: {
        center: [tw.justify_center],
        start: [tw.justify_start],
      },
    },
    defaultVariants: {
      variant: "secondary",
      size: "md",
      block: false,
      disabled: false,
      alignContent: "center",
    },
  },
);

const buttonIcon = tw.h_full.flex.items_center;

export const BaseButton = ({
  trailingIcon,
  trailingAction,
  leadingIcon,
  icon,
  variant,
  size,
  alignContent,
  block,
  disabled,
  children,
  ...props
}: ButtonProps) => {
  return (
    <button
      disabled={disabled}
      className={baseButton({ disabled, variant, size, alignContent, block })}
      {...props}
    >
      {icon ?? (
        <>
          {leadingIcon && <span className={buttonIcon}>{leadingIcon}</span>}
          <span className={tw.overflow_hidden.whitespace_nowrap.text_inherit}>{children}</span>
          {trailingIcon && <span className={buttonIcon}>{trailingIcon}</span>}
          {trailingAction && <span className={buttonIcon}>{trailingAction}</span>}
        </>
      )}
    </button>
  );
};
