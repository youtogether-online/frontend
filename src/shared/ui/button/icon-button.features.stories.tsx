import { type Meta, type StoryObj } from "@storybook/react";

import { Icon } from "../icon";
import { IconButton } from "./icon-button";

const meta: Meta<typeof IconButton> = {
  title: "Components/Buttons/IconButton/Features",
  tags: ["autodocs"],
  component: IconButton,
};

export default meta;
type Story = StoryObj<typeof IconButton>;

export const Primary: Story = {
  args: {
    variant: "primary",
    icon: <Icon name="logos/youtogether-vertical" />,
  },
};

export const Secondary: Story = {
  args: {
    variant: "secondary",
    icon: <Icon name="logos/youtogether-vertical" />,
  },
};

export const Outline: Story = {
  args: {
    variant: "outline",
    icon: <Icon name="logos/youtogether-vertical" />,
  },
};

export const Danger: Story = {
  args: {
    variant: "danger",
    icon: <Icon name="logos/youtogether-vertical" />,
  },
};

export const Small: Story = {
  args: {
    size: "sm",
    icon: <Icon name="logos/youtogether-vertical" />,
  },
};

export const Medium: Story = {
  args: {
    size: "md",
    icon: <Icon name="logos/youtogether-vertical" />,
  },
};

export const Large: Story = {
  args: {
    size: "lg",
    icon: <Icon name="logos/youtogether-vertical" />,
  },
};
