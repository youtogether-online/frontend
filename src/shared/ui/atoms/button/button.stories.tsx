import type { Meta, StoryObj } from "@storybook/react";

import { Button } from "./button";

const meta: Meta<typeof Button> = {
  title: "Components/Buttons/Button/Features",
  component: Button,
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Variants: Story = {
  render: () => (
    <div className="flex gap-4">
      <div className="flex flex-col items-center gap-12">
        <Button variant="primary">Primary</Button>
        <Button variant="secondary">Secondary</Button>
        <Button variant="outline">Outline</Button>
        <Button variant="danger">Danger</Button>
      </div>
      <div className="flex flex-col items-center gap-12">
        <Button variant="primary" disabled>
          Primary
        </Button>
        <Button variant="secondary" disabled>
          Secondary
        </Button>
        <Button variant="outline" disabled>
          Outline
        </Button>
        <Button variant="danger" disabled>
          Danger
        </Button>
      </div>
    </div>
  ),
};

export const Sizes: Story = {
  render: () => (
    <div className="flex items-center gap-12">
      <Button variant="secondary" size="sm">
        Small
      </Button>
      <Button variant="secondary" size="md">
        Medium
      </Button>
      <Button variant="secondary" size="lg">
        Large
      </Button>
    </div>
  ),
};
