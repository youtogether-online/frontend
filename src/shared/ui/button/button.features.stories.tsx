import type { Meta, StoryObj } from "@storybook/react";

import { Icon } from "@/shared/ui";

import { Button } from "./button";

const meta: Meta<typeof Button> = {
  title: "Components/Buttons/Button/Features",
  tags: ["autodocs"],
  component: Button,
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: {
    variant: "primary",
    children: "Primary",
  },
};

export const Secondary: Story = {
  args: {
    variant: "secondary",
    children: "Secondary",
  },
};

export const Outlined: Story = {
  args: {
    variant: "outline",
    children: "Outlined",
  },
};

export const Danger: Story = {
  args: {
    variant: "danger",
    children: "Danger",
  },
};

export const Invisible: Story = {
  args: {
    variant: "invisible",
    children: "Invisible",
  },
};

export const Small: Story = {
  args: {
    size: "sm",
    children: "Small",
  },
};

export const Medium: Story = {
  args: {
    size: "md",
    children: "Medium",
  },
};

export const Large: Story = {
  args: {
    size: "lg",
    children: "Large",
  },
};

export const Disabled: Story = {
  args: {
    disabled: true,
    children: "Disabled",
  },
};

export const LeadingIcon: Story = {
  args: {
    leadingIcon: <Icon name="common/arrowLeft" />,
    children: "Leading icon",
  },
};

export const TrailingIcon: Story = {
  args: {
    trailingAction: <Icon name="common/arrowLeft" />,
    children: "Trailing Icon",
  },
};

export const TrailingAction: Story = {
  args: {
    trailingAction: <Icon name="common/arrowLeft" />,
    children: "Leading Action",
  },
};
