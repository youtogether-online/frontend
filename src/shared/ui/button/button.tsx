import { cva } from "cva";
import { forwardRef } from "react";

import { BaseButton } from "./base-button";
import { type ButtonProps } from "./types";

export const buttonVariants = cva("", {
  variants: {
    size: {
      sm: "text-sm h-7 px-2",
      md: "text-md h-8 px-3",
      lg: "text-md h-10 px-3",
    },
  },
  defaultVariants: {
    size: "md",
  },
});

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ children, size, sx, ...props }, ref) => {
    return (
      <BaseButton ref={ref} sx={buttonVariants({ size, class: sx })} {...props}>
        {children}
      </BaseButton>
    );
  },
);
