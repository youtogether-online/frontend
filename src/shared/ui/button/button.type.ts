import { type VariantProps } from "cva";
import { type ButtonHTMLAttributes, type ReactNode } from "react";

import { type baseButton } from "./base-button";

export type BaseButtonProps = VariantProps<typeof baseButton> &
  ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = {
  trailingIcon?: ReactNode;
  trailingAction?: ReactNode;
  leadingIcon?: ReactNode;
  icon?: ReactNode;
} & BaseButtonProps;

export type IconButtonProps = {
  icon: ReactNode;
} & BaseButtonProps;
