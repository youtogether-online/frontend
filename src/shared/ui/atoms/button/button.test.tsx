import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";

import { Button } from "./button";

describe("Button", () => {
  it("Button should respects block prop", () => {
    const container = render(<Button block>Block</Button>);
    const button = container.getByRole("button");
    expect(button).toMatchSnapshot();
  });
});
