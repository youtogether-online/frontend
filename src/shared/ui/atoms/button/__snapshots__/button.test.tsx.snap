// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > Button should respects block prop 1`] = `
<button
  class="transition-colors border border-solid rounded-md cursor-pointer bg-buttonBg text-buttonText border-buttonBorder hover:border-buttonHoverBorder hover:bg-buttonHoverBg active:bg-buttonBg active:text-buttonText active:border-buttonBorder disabled:text-fgMuted text-md h-8 px-3 w-full"
>
  <span
    class="flex h-full items-center overflow-hidden whitespace-nowrap"
  >
    Block
  </span>
</button>
`;
